api/src/main/java/org/apache/iceberg/transforms/Bucket.java

```diff
@@ -90,6 +96,11 @@ public Integer apply(T value) {
     return (hash(value) & Integer.MAX_VALUE) % numBuckets;
   }
 
+  @Override
+  public boolean canTransform(Type type) {
+    return type.isPrimitiveType();
```

Comment:
is this correct? If we can `canTransform(StringType)` on `BucketLong`, this will pass. Or caller always use the same `type` as the bucket type and this is not a concern?

Reply:
It seems like a valid point cause we still allow to construct `Bucket` for a particular type. Should we deprecate/prohibit that?

```
static <T> Bucket<T> get(Type type, int numBuckets) {
```

Reply:
I think this is correct because it is whether the transform can be applied to a type, not whether the current instance of that transform can be applied to a type. This is used to validate partition specs, sort orders, and bound functions that contain transforms. When we know the concrete type, we check whether the transform can handle it.

Before, we would get the final transform at the point just before that check, when the type was known. That was basically using `Transform.fromString(actualType, transform.toString())` to rebuild the transform and then it would call `canTransform`. Now the same generic transform can be used and `canTransform` doesn't need a specific transform instance.